tree/30_generic_methods/file_line_present_in_ini_section.cf

```diff
@@ -34,32 +34,60 @@ bundle agent file_line_present_in_ini_section(file, section, line)
       "report_param"           string => join("_", args);
       "full_class_prefix"      string => canonify("file_line_present_in_ini_section_${report_param}");
       "class_prefix"           string => string_head("${full_class_prefix}", "1000");
-      "section_and_blank_line" string => "[${section}]
-";
+      "hash"                   string => hash("${report_param}", "md5");
+      "last_line"              string => execresult("/usr/bin/tail -n 1 ${file} #${hash}", "useshell");
+
+    pass2::
+      "line_list" slist =>  splitstring("${line}", "${const.n}", "999999");
+      "unique_lines" slist =>  unique("line_list");
```

Comment:
i don't understand why this is deduplicated. nowhere in the method it is stated that it should behave like that, and this is really surprising for user

Reply:
This is the current behaviour, it currently only uses "insert_lines" promises so you can't have 2 times the same line in your section. And here it done manually since the section delimiters are limited in cfengine and do not work if the section is the last one in a file and is empty, the section automatically becomes the whole file.... so we need to create a block of text, clean its duplicated lines and insert it at the end without any duplication check.

Reply:
it was ncf_insert_block, which is insert_type => "preserve_block", so it doesn't deduplicate

Reply:
This was an issue: at section creation it used ncf_insert_block, with preserver_block (which in fact still do a dedupolicate but with the whole block and not line per line), but afterward, when the section exist, it uses "ensure_line_in_ini_section" [https://github.com/Normation/ncf/blob/fc517e756083b0d05f44001d17f01e785b3a2f27/tree/30_generic_methods/file_line_present_in_ini_section.cf#L57](url)

Reply:
Which is defined here: [https://github.com/Normation/ncf/blob/fc517e756083b0d05f44001d17f01e785b3a2f27/tree/20_cfe_basics/files.cf#L232](url)

Reply:
And does apply a "deduplicate"